feat: allow custom compileQuery implementation

hoangvvo opened the pull request:
This PR allows `@benzene/core` to have customizable `compileQuery` function. This allows us to either use `graphql-js` or `graphql-jit`.

doc: https://benzene-git-feat-custom-compilequery-hoangvvo.vercel.app/reference/runtime

Related issues: #48 

Benchmarks:

```
Machine: Linux 5.11.0-17-generic x86_64 | 12 vCPUs | 16GB
Node: v16.0.0

apollo-server
  Slowest:	0.0631 secs
  Fastest:	0.0051 secs
  Average:	0.0309 secs
  Requests/sec:	3228.3084
  
benzene-http
  Slowest:	0.0275 secs
  Fastest:	0.0052 secs
  Average:	0.0089 secs
  Requests/sec:	11180.8484
  
benzene-jit-http
  Slowest:	0.0171 secs
  Fastest:	0.0023 secs
  Average:	0.0041 secs
  Requests/sec:	24212.8719
  
express-graphql
  Slowest:	0.0815 secs
  Fastest:	0.0052 secs
  Average:	0.0377 secs
  Requests/sec:	2644.2127
  
graphql-helix
  Slowest:	0.0778 secs
  Fastest:	0.0058 secs
  Average:	0.0413 secs
  Requests/sec:	2413.2614
  
mercurius
  Slowest:	0.0176 secs
  Fastest:	0.0029 secs
  Average:	0.0047 secs
  Requests/sec:	21452.4651
```

A breaking change is that the default will be switched to `graphql-js`.

hoangvvo commented on docs/pages/reference/runtimes.md:
```diff
@@ -0,0 +1,116 @@
+# Runtimes
```
Is this a good term for it? I am also considering "Executors", "Engines", ...

rfgamaral replied:
Of the three, I prefer `Runtimes`, just like you ahve it now :)

hoangvvo commented:
I am also considering moving `@benzene/jit` to a seperate package. Not sure if this will cause much inconvenience.

rfgamaral commented:
I have not yet tested this code, but a few observations:

1. It looks to me like the documentation starts with explaining how one can customize the runtime, and only then mentions the available built-in implementations, correct? I'd probably switch that, first mention the available built-in runtimes, and only then provide advanced documentation for a custom implementation. I was confused when I started reading the documentation because all I seem to need to run this with Cloudflare Workers is to import he correct `compileQuery` and use it.
2. I'd probably suggest a bigger introduction to runtimes, what they are for, and why they are useful. Perhaps mentioning the Cloudflare Workers example where it can't use `eval`, and that's why it needs a different runtime. But if users are running this on a different platform that doesn't have this limitation, one can use the more performant runtime.
3. > I am also considering moving `graphql-jit` implementation to a separate package. Not sure if this will cause much inconvenience.

   As long as that is properly documented, I'd say go for it. I'd probably drop the "runtime" from the package name though. I understand why you added it, but if in the future you feel like "runtimes" is not the best name, changing the package name is definitely an inconvenience for everyone (you and users). This makes the imports more clean for users, and it's probably easier for you to maintain.
4. I might have some documentation nitpicks, but I'll leave it to a later time in case you want to follow up on my suggestions on point 1 and 2.

rfgamaral commented:
> benzene-http
>   Slowest:	0.0275 secs
>   Fastest:	0.0052 secs
>   Average:	0.0089 secs
>   Requests/sec:	11180.8484
>   
> benzene-jit-http
>   Slowest:	0.0171 secs
>   Fastest:	0.0023 secs
>   Average:	0.0041 secs
>   Requests/sec:	24212.8719

Wow, that's a huge difference... Too bad I can't use the "jit" implementation, if only Cloudflare Workers supported `eval` things. There's really no way to maybe improve `graphql-jit` in a way that doesn't use `eval` things? Maybe in a future version with more time to work on that?

> A breaking change is that the default will be switched to `graphql-js`.

May I suggest making a small suggestion in the "getting started" page for users to look into use switching to  `graphql-jit` if they can, and take advantage of the double the requests/sec? I'm afraid that most people don't read the whole documentation and might miss that performance improvement that maybe there platform allows for.

hoangvvo commented:
> > benzene-http
> > Slowest:	0.0275 secs
> > Fastest:	0.0052 secs
> > Average:	0.0089 secs
> > Requests/sec:	11180.8484
> > benzene-jit-http
> > Slowest:	0.0171 secs
> > Fastest:	0.0023 secs
> > Average:	0.0041 secs
> > Requests/sec:	24212.8719
> 
> Wow, that's a huge difference... Too bad I can't use the "jit" implementation, if only Cloudflare Workers supported `eval` things. There's really no way to maybe improve `graphql-jit` in a way that doesn't use `eval` things? Maybe in a future version with more time to work on that?
> 
> > A breaking change is that the default will be switched to `graphql-js`.
> 
> May I suggest making a small suggestion in the "getting started" page for users to look into use switching to `graphql-jit` if they can, and take advantage of the double the requests/sec? I'm afraid that most people don't read the whole documentation and might miss that performance improvement that maybe there platform allows for.

@rfgamaral  Huge difference but still much better than others. There is another way though and that is to compile the query before deploying to Cloudflare Worker. That way we can still use jit without eval. One limitation is that you need to know your queries before hand.

hoangvvo commented:
Thanks for all the comments. They are truly useful! I will get some commits in.

hoangvvo commented:
> I have not yet tested this code, but a few observations:
> 
> 1. It looks to me like the documentation starts with explaining how one can customize the runtime, and only then mentions the available built-in implementations, correct? I'd probably switch that, first mention the available built-in runtimes, and only then provide advanced documentation for a custom implementation. I was confused when I started reading the documentation because all I seem to need to run this with Cloudflare Workers is to import he correct `compileQuery` and use it.
> 2. I'd probably suggest a bigger introduction to runtimes, what they are for, and why they are useful. Perhaps mentioning the Cloudflare Workers example where it can't use `eval`, and that's why it needs a different runtime. But if users are running this on a different platform that doesn't have this limitation, one can use the more performant runtime.
> 3. > I am also considering moving `graphql-jit` implementation to a separate package. Not sure if this will cause much inconvenience.
>    
>    
>    As long as that is properly documented, I'd say go for it. I'd probably drop the "runtime" from the package name though. I understand why you added it, but if in the future you feel like "runtimes" is not the best name, changing the package name is definitely an inconvenience for everyone (you and users). This makes the imports more clean for users, and it's probably easier for you to maintain.
> 4. I might have some documentation nitpicks, but I'll leave it to a later time in case you want to follow up on my suggestions on point 1 and 2.

@rfgamaral  Thank you. I made the following adjustment.

1. I moved the `Built-in implementations` to above. I also make a note about the limitation of environment like Cloudflare.
2. Added a reference as a "feature" and also add a reference to Getting Started, along with benchmark statistics reference.
3. For now I will just bundle it with the codebase.
4. let me know any nitpicks you can find.

Thanks again!

rfgamaral commented on docs/pages/reference/runtime.md:
```diff
@@ -0,0 +1,131 @@
+
+**Benzene** supports GraphQL execution using a customizable runtime.
+
+Benzene's GraphQL runtime can be configured using the `compileQuery` option. By default, it will use [`graphql-js` implementation](#graphql-js).
```
```suggestion
Benzene's GraphQL runtime can be configured with the `compileQuery` option. By default, it will use [`graphql-js`](#graphql-js) implementation.
```

rfgamaral commented on docs/pages/reference/runtime.md:
```diff
@@ -0,0 +1,131 @@
+
+## Built-in implementations
+
+`@benzene/core` is bundled two of the implementations, [`graphql-js`](https://github.com/graphql/graphql-js) and [`graphql-jit`](https://github.com/zalando-incubator/graphql-jit).
```
```suggestion
`@benzene/core` currently bundles two implementations, [`graphql-js`](https://github.com/graphql/graphql-js) and [`graphql-jit`](https://github.com/zalando-incubator/graphql-jit).
```

rfgamaral commented on docs/pages/reference/runtime.md:
```diff
@@ -0,0 +1,131 @@
+
+`@benzene/core` is bundled two of the implementations, [`graphql-js`](https://github.com/graphql/graphql-js) and [`graphql-jit`](https://github.com/zalando-incubator/graphql-jit).
+
+The modules are ESModule bundled, so the importing process may be different from the usual:
```
```suggestion
The modules are ESModule bundled, so the importing process may be different than usual:
```

rfgamaral commented on docs/pages/reference/runtime.md:
````diff
@@ -0,0 +1,131 @@
+});
+```
+
+This is also preferred if your application is deployed in specific environments, such as [Cloudflare Worker](https://workers.cloudflare.com/), which does not allow runtime evaluation required by other implementations.
````
```suggestion
This is also preferred if your application is deployed in specific environments, such as [Cloudflare Workers](https://workers.cloudflare.com/), which does not allow runtime evaluation required by other implementations.
```

rfgamaral commented on docs/pages/reference/runtime.md:
````diff
@@ -0,0 +1,131 @@
+});
+```
+
+Check out [benchmarks](/benchmarks) so see how using this runtime can benefit your application.
````
```suggestion
Check out the [benchmarks](/benchmarks) so see how using this runtime can benefit your application.
```

rfgamaral commented on docs/pages/reference/runtime.md:
```diff
@@ -0,0 +1,131 @@
+
+The function will be called with the schema, document node, and optionally operation name and returns either:
+
+- an object with three functions `execute`, `subscribe`, and optionally `stringify`.
```
```suggestion
- An object with three functions `execute`, `subscribe`, and optionally `stringify`.
```

rfgamaral commented on docs/pages/reference/runtime.md:
```diff
@@ -0,0 +1,131 @@
+- an object with three functions `execute`, `subscribe`, and optionally `stringify`.
+- GraphQL execution result in case of error
+
+The signatures of `execute` and `subscribe` are the same as those (non-positional ones) from `graphql-js` (A difference being that it does not contain `schema` in the argument).
```
```suggestion
The signatures of `execute` and `subscribe` are the same as those (non-positional ones) from `graphql-js` (a difference being that it does not contain `schema` in the argument).
```

rfgamaral reviewed (commented):
@hoangvvo Left a few documentation nitpicks...

Also tried to get this running locally, so I could test it with Cloudflare Workers but I couldn't get it to work. Here's how I tried:

1. Cloned this repository, and checked out `feat/custom-compileQuery`
2. Ran `yarn install` and `yarn run build`
3. Changed to `packages/core` and ran `npm link`
4. Changed to `packages/http` and ran `npm link`
5. Tried to run my basic server with a simple query, got this:
    ```
    Uncaught
    Error: Cannot use GraphQLSchema2 "[object GraphQLSchema]" from another module or realm.

    Ensure that there is only one instance of "graphql" in the node_modules
    directory. If different versions of "graphql" are the dependencies of other
    relied on modules, use "resolutions" to ensure only one version is installed.

    https://yarnpkg.com/en/docs/selective-version-resolutions

    Duplicate "graphql" modules cannot be used at the same time since different
    versions may have different capabilities and behavior. The data from one
    version used in the function from another could produce confusing and
    spurious results.
        at instanceOf (worker.js:853:17)
        at isSchema (worker.js:5294:38)
        at assertSchema (worker.js:5297:12)
        at validateSchema2 (worker.js:5532:32)
        at new Benzene (worker.js:35281:70)
        at worker.js:35468:11
    Uncaught (in response)
    Error: Cannot use GraphQLSchema2 "[object GraphQLSchema]" from another module or realm.

    Ensure that there is only one instance of "graphql" in the node_modules
    directory. If different versions of "graphql" are the dependencies of other
    relied on modules, use "resolutions" to ensure only one version is installed.

    https://yarnpkg.com/en/docs/selective-version-resolutions

    Duplicate "graphql" modules cannot be used at the same time since different
    versions may have different capabilities and behavior. The data from one
    version used in the function from another could produce confusing and
    spurious results.
    ```

Not sure what I'm missing... Maybe you could publish new "core" and "http" versions with a `preview` [dist-tag](https://docs.npmjs.com/cli/v7/commands/npm-dist-tag)?

hoangvvo commented:
> @hoangvvo Left a few documentation nitpicks...
> 
> Also tried to get this running locally, so I could test it with Cloudflare Workers but I couldn't get it to work. Here's how I tried:
> 
> 1. Cloned this repository, and checked out `feat/custom-compileQuery`
> 2. Ran `yarn install` and `yarn run build`
> 3. Changed to `packages/core` and ran `npm link`
> 4. Changed to `packages/http` and ran `npm link`
> 5. Tried to run my basic server with a simple query, got this:
>    ```
>    Uncaught
>    Error: Cannot use GraphQLSchema2 "[object GraphQLSchema]" from another module or realm.
>    
>    Ensure that there is only one instance of "graphql" in the node_modules
>    directory. If different versions of "graphql" are the dependencies of other
>    relied on modules, use "resolutions" to ensure only one version is installed.
>    
>    https://yarnpkg.com/en/docs/selective-version-resolutions
>    
>    Duplicate "graphql" modules cannot be used at the same time since different
>    versions may have different capabilities and behavior. The data from one
>    version used in the function from another could produce confusing and
>    spurious results.
>        at instanceOf (worker.js:853:17)
>        at isSchema (worker.js:5294:38)
>        at assertSchema (worker.js:5297:12)
>        at validateSchema2 (worker.js:5532:32)
>        at new Benzene (worker.js:35281:70)
>        at worker.js:35468:11
>    Uncaught (in response)
>    Error: Cannot use GraphQLSchema2 "[object GraphQLSchema]" from another module or realm.
>    
>    Ensure that there is only one instance of "graphql" in the node_modules
>    directory. If different versions of "graphql" are the dependencies of other
>    relied on modules, use "resolutions" to ensure only one version is installed.
>    
>    https://yarnpkg.com/en/docs/selective-version-resolutions
>    
>    Duplicate "graphql" modules cannot be used at the same time since different
>    versions may have different capabilities and behavior. The data from one
>    version used in the function from another could produce confusing and
>    spurious results.
>    ```
> 
> Not sure what I'm missing... Maybe you could publish new "core" and "http" versions with a `preview` [dist-tag](https://docs.npmjs.com/cli/v7/commands/npm-dist-tag)?

This is due to two `graphql-js` being installed in `node_modules` or nested `node_modules`. I will try to publish a beta tag.

hoangvvo commented:
@rfgamaral I published the packages as `next`. Try it out!

```
npm i @benzene/http@next
```

rfgamaral commented:
> 
> 
> @rfgamaral I published the packages as `next`. Try it out!
> 
> ```
> npm i @benzene/http@next
> ```

@hoangvvo It's working 🎉

I'm only having one small issue... I can compile and run, but my ESLint rules are giving me this:

![image](https://user-images.githubusercontent.com/96476/119516035-4e248300-bd6e-11eb-9a5b-4d837bf58691.png)

If I change the import to `@benzene/core/dist/runtimes/js`, the error goes away, and I get a typed `makeCompileQuery`, but now my build script fails.

Is there anyway we can solve this **without** adding ignoring the ESLint error? 🤔

hoangvvo commented:
It seems that TypeScript does not play well with ESModule for now. I may have to rethink the import/export approach.

rfgamaral commented:
> 
> 
> What is the build fail error? @rfgamaral

If I change the import to `@benzene/core/dist/runtimes/js`, this is what I get when I try to run my server:

```
 > src/core/benzene.ts:1:33: error: Could not resolve "@benzene/core/dist/runtimes/js" (mark it as external to exclude it from the bundle)
     1 │ import { makeCompileQuery } from '@benzene/core/dist/runtimes/js'
       ╵                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   node_modules/@benzene/core/package.json:20:13: note: The path "./dist/runtimes/js" is not exported by package "@benzene/core"
    20 │   "exports": {
       ╵              ^
   node_modules/@benzene/core/package.json:30:16: note: The file "./dist/runtimes/js.js" is exported at path "./runtimes/js"
    30 │       "import": "./dist/runtimes/js.js",
       ╵                 ~~~~~~~~~~~~~~~~~~~~~~~
   src/core/benzene.ts:1:33: note: Import from "@benzene/core/runtimes/js" to get the file "node_modules/@benzene/core/dist/runtimes/js.js"
     1 │ import { makeCompileQuery } from '@benzene/core/dist/runtimes/js'
       │                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       ╵                                  "@benzene/core/runtimes/js"

Error: Build failed! Status Code: 1
```

hoangvvo commented:
Are you using a build tool? If so, which one?

rfgamaral commented:
> 
> 
> @rfgamaral Are you using a build/bundle tool? If so, which one?

Yes, I'm using `esbuild`, here's the `build.ts` script I use:

```ts
#!/usr/bin/env ts-node-script

import type { BuildOptions } from 'esbuild'

const isDevelopment = process.env.NODE_ENV === 'development'
const isProduction = process.env.NODE_ENV === 'production'

const buildOptions: BuildOptions = {
    entryPoints: ['./src/worker.ts'],
    mainFields: ['main'],
    platform: 'neutral',
    define: {
        'process.env.NODE_ENV': `"${process.env.NODE_ENV}"`,
    },
    bundle: true,
    minify: isProduction,
    sourcemap: isDevelopment,
    outfile: './dist/worker.js',
}

import('esbuild').then(({ build }) =>
    build(buildOptions).catch(() => {
        process.exit(1)
    })
)
```

hoangvvo commented:
@rfgamaral Can you please try the next next version?

```
npm i @benzene/core@0.6.0-next.1 @benzene/http@0.2.0-next.1
```

hoangvvo commented:
@rfgamaral Ooops, 1 possible problem. If you have a problem with `require() of x from y is an ES module file`, please try patch the file `node_modules/@benzene/core/index.cjs` to replace `require('./runtimes/js')` with `require('./runtimes/js.cjs')`

rfgamaral commented:
> 
> 
> @rfgamaral Can you please try the next next version?
> 
> ```
> npm i @benzene/core@0.6.0-next.1 @benzene/http@0.2.0-next.1
> ```

Changed my imports to:

```ts
import { makeCompileQuery } from '@benzene/core/runtimes/js'
import { Benzene, parseGraphQLBody, makeHandler } from '@benzene/http'
import type { HTTPRequest, HTTPResponse } from '@benzene/http/types'
```

But now I got this:

```
> src/core/benzene.ts:2:55: error: Could not resolve "@benzene/http" (mark it as external to exclude it from the bundle)
     2 │ import { Benzene, parseGraphQLBody, makeHandler } from '@benzene/http'
       ╵                                                        ~~~~~~~~~~~~~~~
   node_modules/@benzene/http/package.json:27:13: note: The path "." is not currently exported by package "@benzene/http"
    27 │   "exports": {
       ╵              ^
   node_modules/@benzene/http/package.json:28:9: note: None of the conditions provided ("export", "require") match any of the currently active conditions ("default", "import")
    28 │     ".": {
       ╵          ^
   src/core/benzene.ts:2:55: note: Consider using a "require()" call to import this file
     2 │ import { Benzene, parseGraphQLBody, makeHandler } from '@benzene/http'
       ╵                                                        ~~~~~~~~~~~~~~~

Error: Build failed! Status Code: 1
```

Which is weird because ESLint is not complaining, and the types show up correctly on VS Code.

hoangvvo commented:
Ooops, can you manually patch `node_modules/@benzene/http/package.json`, replace `export` with `import`?

![image](https://user-images.githubusercontent.com/40987398/119545058-8ce12a00-bdbc-11eb-9082-9c4db6f8e019.png)

hoangvvo commented:
@rfgamaral I guess we might need to consider moving the `jit` functionality away from this package, to allow `makeCompileQuery` to imported from root (`import { makeCompileQuery } from "@benzene/core"`) since importing nested file seems very troubling for ESModule atm.

Maybe the new package can be `@benzene/jit`?

What do you think?

rfgamaral commented:
> 
> 
> @rfgamaral Ooops, can you manually patch `node_modules/@benzene/http/package.json`, replace `export` with `import`?
> 
> ![image](https://user-images.githubusercontent.com/40987398/119545058-8ce12a00-bdbc-11eb-9082-9c4db6f8e019.png)

This worked 👍

> Maybe the new package can be `@benzene/jit`? Moving it away can also helps lowering the size of `@benzene/core` significantly.
> 
> What do you think?

IS this `@benzene/jit` package meant for both runtimes, or are you keeping `graphql-js` built-in as the default? If the former (both packages), I'd probably use `@benzene/runtimes` to be consistent with the documentation. If the latter, looks like a good package name to me.

I actually like the idea of having Benzene work with `graphql-js` by default without needing any dependencies other than `@benzene/http`, and that if you want/need more performance, you can just add `@benzene/jit` and use it. If you go this route, maybe the message we get about explicitly setting the `compileQuery` is no longer needed? 🤔

hoangvvo commented:
@rfgamaral `@benzene/jit` will contain only the graphql-jit implementation. The default graphql-js will be bundled in `@benzene/core`. 

Without it, we can look forward to shaving off 45.54kb from the bundle (so we are only left with only ~6kb for this package, truly lightweight!), which makes it works pretty well for env like serverless or worker.

hoangvvo commented:
I think the message should stay because `graphql-jit` used to be the default, which makes this change a breaking change. We need to notify those who take advantages of `graphql-jit` in the past to not miss out and get a perf penalty without knowledge.

rfgamaral commented:
> 
> 
> @rfgamaral `@benzene/jit` will contain only the graphql-jit implementation. The default graphql-js will be bundled in `@benzene/core`.
> 
> Without having `graphql-jit` bundled in `@benzene/core`, we can look forward to shaving off 45.54kb from the bundle (so we are only left with only ~6kb for this package, truly lightweight!), which makes it works pretty well for env like serverless or worker.

Makes sense to me, go for it 👍

hoangvvo commented:
This PR is quite big. I will merge it for now and will follow up with a PR to clean everything up. For now we will go with the approach to have `@benzene/jit` as a seperate package.

hoangvvo commented:
I have released `@benzene/core@0.5.0`, `@benzene/http@0.2.0`, `@benzene/jit@0.1.0` among others!

@rfgamaral

rfgamaral commented:
Awesome work @hoangvvo, thank you very much for this :)
